content/6_Advanced/String_Search.mdx

```diff
@@ -152,19 +150,124 @@ but instead finds the longest common prefix of a string and all of its suffixes.
 **Manacher's Algorithm** functions similarly to the **Z-Algorithm**. It
 determines the longest palindrome centered at each character.
 
+Let's denote $dp_i$ as the maximum diameter of a palindrome centered at $i$.
+Manacher's algorithm makes use of the previously determined $dp_j$, where $j < i$ incalculating $dp_i$. The main idea is that for a palindrome centered at $i$ with the borders $left$ and $right$ the $dp_j$ ($i < j \le right$) values are - probably - mirrors
+of the $dp_k$ ($left \le k < i$) values on the left side of the palindrome.
+Probably because for some $j$ the maximum palindrome might cross the right border.
+This way the algorithm only considers the palindrome centers that could lead to the expansion of the maximum palindrome.
+
+**Time complexity:** $\mathcal{O}(N)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+string menacher(string s) {
+	// Preprocess the input so it can handle even length palindromes
+	string arr;
+	for (int i = 0; i < s.size(); i++) {
+		arr.push_back('#');
+		arr.push_back(s[i]);
+	}
+	arr.push_back('#');
+
+	// dp[i] = palindrome's maximum diameter centered at i
+	vector<int> dp(arr.size());
+	int left = 0;
+	int right = 0;
+	int lg_max = 0;
+	int idx = 0;
+	for (int i = 0; i < arr.size();) {
+		// Expand the palindrome around i
+		while (left > 0 && right < arr.size() - 1 &&
+		       arr[left - 1] == arr[right + 1]) {
+			left--;
+			right++;
+		}
+
+		// Update the diameter
+		dp[i] = right - left + 1;
+
+		if (lg_max < dp[i]) {
+			lg_max = dp[i];
+			idx = i;
+		}
+
+		/*
+		 * We've already computed the values for [left, i] interval.
+		 * The values on the right side of the palindrome [i+1, right] will
+		 * probably be the same as the ones on the left side. Probably because
+		 * some might even go beyond the current [left, right] palindrome.
```

Comment:
wdym "probably"

Reply:
I said right there: `Probably  because some might even go beyond the current [left, right] palindrome`

Reply:
Or should I write: `I said probably because some might even go beyond the current [left, right] palindrome`.

Reply:
ftfy
also what is "some" referring to

Reply:
It says right there: `Probably because for some $j$ the maximum palindrome might cross the right border.
`. The comment preview is outdated.

Reply:
<img width="382" alt="image" src="https://github.com/cpinitiative/usaco-guide/assets/55369003/413a67ca-05fb-4c14-a438-f85b2bb021a0">

uh i don't see any mention of $j$ here in this comment?

Reply:
I added `palindromes`